### ISSUE TITLE
Type warning disappears after using `reveal_type` (+ false positive?)

### ISSUE THREAD

**Issue by bram-tv** [microsoft/pyright#8090]
**Describe the bug**

Using the following example code (stripped down from a large code base[^1]):

```python
# pyright: strict

from typing import Optional


class Example:
    def _calculate_changes(self) -> None:
        self.changes: Optional[int] = 0
        for _ in range(1, 10):
            self.changes += 1
            # reveal_type(self.changes)
```

and running pyright on it:
```
.../example.py
  .../example.py:10:13 - error: Type of "changes" is partially unknown
    Type of "changes" is "Unknown | int | None" (reportUnknownMemberType)
  .../example.py:10:13 - error: Operator "+=" not supported for types "Unknown | int | None" and "Literal[1]"
    Operator "+" not supported for types "None" and "Literal[1]" (reportOperatorIssue)
2 errors, 0 warnings, 0 informations
```

Running again after uncommenting the `reveal_type` line:
```
.../example.py
  .../example.py:11:25 - information: Type of "self.changes" is "int"
0 errors, 0 warnings, 1 information
```

There are two potential bugs:
- Error that disappears after using `reveal_type()`
- False positive for the above code(?) (mypy does not detect an error in the code)

**VS Code extension or command-line**

Using pyright 1.1.366

[^1]: In the example code the type hint can be changed to `self.changes: int` and it will work but in the larger code base the variable is initialized to `None` in the `__init__` method which is why it is typed as optional.

**Comment by LeeeeT** [comment]
You can, in fact, reveal the type of any expression using `reveal_type` (not specifically `self.changes`) and it will mess with the type of `self.changes` for some reason.

```python
class Example:
    def _calculate_changes(self) -> None:
        self.changes: Optional[int] = 0
        for _ in range(1, 10):
            self.changes += 1  # no error
            reveal_type(1)
```

Using `reveal_type` somehow causes pyright to narrow the type of `self.changes` to `int` despite the annotation.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.367.
